### ISSUE TITLE
How to get or see xgboost's gradient statistics value?

### ISSUE THREAD

**Issue by bughunter99** [dmlc/xgboost#2489]
I'm studying xgboost and novice for gradient boost. In gradient tree boosting, loss function is derived by second order approximation calculating gi, hi. You can see it on https://xgboost.readthedocs.io/en/latest/model.html#the-structure-score . Given a dataset, how can I see the value gi, hi such as g1, h1, g2, h2,..?

I saw _train_internal and several functions in training.py and sklean.py. But I didn't find it. By understanding how it calculated and got efficiently, it could be possible to apply further algorithm used in xgboost such as quantile percentile sketch.

Thanks.

**Comment by hcho3** [comment]
The gradient computation occurs in the backend, which is written in C++. Each objective class implements a function `GetGradient` that computes gradient statistics given predicted and true labels. See [here](https://github.com/dmlc/xgboost/blob/master/src/objective/regression_obj.cc) for example.

**Comment by bughunter99** [comment]
Thank you, But I still could not understand what G, H are in objective function in xgboost. Of course I understood how the equation is derivated. In there, they are first and second gradient from taylor series,.. May I know what they is and what meaning they have intuitively?

**Comment by hcho3** [comment]
Each data point has a gradient pair (g_i, h_i), which is given as follows:
```
g_i = Loss::FirstOrderGradient([predicted label for x_i], [true label for x_i])
h_i = Loss::SecondOrderGradient([predicted label for x_i], [true label for x_i])
```
The variables (G_j, H_j) denote the sum of gradients over all data points in the j-th instance set.

**Comment by bughunter99** [comment]
Thank you, I think the reason why I don't understand well is from not understanding what gradient is fully and exactly. What I have been known about gradient is given a function such as objective function or loss function, calculating a slope at certain position. But in xgboost paper, gradient g_i and h_i are included in loss function that is used for score. Then it means given a dataset x1, x2,.. xi, is it possible to calculate g1, h1, g2, h2... gi, hi? I'm little bit confused.

**Comment by hcho3** [comment]
It's best to think of gradient pairs as partial derivatives of the individual loss function with respect to the predicted label. Given the **individual** loss function 
![image](https://user-images.githubusercontent.com/2532981/28498330-7adb8a86-6f4f-11e7-904f-82c00cce2523.png)
which is a function of two variables (true and predicted labels), we compute the first and second partial derivative with respect to the predicted label (second argument below):
![image](https://user-images.githubusercontent.com/2532981/28498336-a98b5410-6f4f-11e7-8d08-bde2774f6bf9.png)
![image](https://user-images.githubusercontent.com/2532981/28498337-b087355e-6f4f-11e7-97c3-445696fda5be.png)

With some algebra, we can show that the **global** loss function (one we use to compute optimal splits) can be computed in terms of gradient pairs.

**Comment by bughunter99** [comment]
Thank you very much. I thought I almost understood, and below stuffs were helpful.

https://stats.stackexchange.com/questions/202858/loss-function-approximation-with-taylor-expansion

https://stats.stackexchange.com/questions/291377/splitting-trees-on-exponential-loss-function
